execution-service/src/main/java/org/ihtsdo/rvf/execution/service/MysqlFailuresExtractor.java

```diff
@@ -0,0 +1,187 @@
+        for (TestRunItem item : items) {
+            String key = String.valueOf(uuidToAssertionIdMap.get(item.getAssertionUuid()));
+            if (assertionIdToTotalFailureMap.containsKey(key)) {
+                item.setFailureCount(Long.valueOf(assertionIdToTotalFailureMap.get(key)));
+                item.setFirstNInstances(fetchFailureDetails(connection, config.getExecutionId(), uuidToAssertionIdMap.get(item.getAssertionUuid()), config.getFailureExportMax(), null, null));
+            } else {
+                item.setFailureCount(0L);
+                item.setFirstNInstances(null);
+            }
+        }
+    }
+
+    private void validateFailuresAndExtractTestResults(Connection connection, List<TestRunItem> items, MysqlExecutionConfig config, Map<UUID, Long> uuidToAssertionIdMap, Map<String, Integer> assertionIdToTotalFailureMap) throws SQLException {
+        for (TestRunItem item : items) {
+            String key = String.valueOf(uuidToAssertionIdMap.get(item.getAssertionUuid()));
+            if (assertionIdToTotalFailureMap.containsKey(key)) {
+                int batch_counter = 0;
+                int total_failures_extracted = 0;
+                int total_whitelistedItem_extracted = 0;
+                int total_failures = assertionIdToTotalFailureMap.get(key);
+                List<FailureDetail> firstNInstances = new ArrayList<>();
+                while(batch_counter * config.getFailureExportMax() < total_failures && total_failures_extracted < config.getFailureExportMax()) {
+                    int startIndex = batch_counter * config.getFailureExportMax();
+                    int endIndex = startIndex + config.getFailureExportMax();
+                    List<FailureDetail> failureDetails = fetchFailureDetails(connection, config.getExecutionId(), uuidToAssertionIdMap.get(item.getAssertionUuid()), -1, startIndex, endIndex);
+
+                    // Convert to WhitelistItem
+                    List<WhitelistItem> whitelistItems = failureDetails.stream()
+                            .map(failureDetail -> new WhitelistItem(key, StringUtils.isEmpty(failureDetail.getComponentId())? "" : failureDetail.getComponentId(), failureDetail.getConceptId(), getAdditionalFields(connection,failureDetail)))
+                            .collect(Collectors.toList());
+
+                    // Send to Authoring acceptance gateway
+                    List<WhitelistItem> whitelistedItems = whitelistService.validateAssertions(whitelistItems);
+
+                    // Find the failures which are not in the whitelisted item
+                    List<FailureDetail> noneWhitelistedFailures = failureDetails.stream().filter(failureDetail ->
+                        whitelistedItems.stream().noneMatch(whitelistedItem ->
+                                failureDetail.getConceptId().equals(whitelistedItem.getConceptId())
```

Comment:
Is conceptId always going to be available? Just using the componentId would be fine.